File: frontend/src/app/shared/models/book-item.ts
import { BookItemStatus } from "../enums/book-item-status";
import { Book } from "./book";

import { Rack, WarehouseItem } from "./rack";

export class BookItem {
    id: number;
    barcode: string;
    isReferenceOnly: boolean;
    borrowed: Date;
    dueDate: Date;
    price: BigInt;
    status: BookItemStatus;
    dateOfPurchase: Date;
    bookId: number;
    rackId: number;
    shelfId: number;
}

export class BookItemWithBook implements WarehouseItem {
    id: number;
    barcode: string;
    isReferenceOnly: boolean;
    borrowed: Date;
    dueDate: Date;
    price: BigInt;
    status: BookItemStatus;
    dateOfPurchase: Date;
    book: Book;
    rackId: number;
    rackName: string;
    shelfId: number;
    shelfName: string;
}

File: backend/old-common/src/main/java/com/example/libraryapp/domain/bookitem/model/BookItem.java
package com.example.libraryapp.domain.bookitem.model;

import com.example.libraryapp.domain.book.model.BookId;
import com.example.libraryapp.domain.bookitemloan.model.LoanCreationDate;
import com.example.libraryapp.domain.bookitemloan.model.LoanDueDate;
import com.example.libraryapp.domain.rack.model.RackId;
import com.example.libraryapp.domain.shelf.model.ShelfId;
import lombok.*;

@Getter
@Setter
@Builder
@AllArgsConstructor
@RequiredArgsConstructor
public class BookItem {
    private BookItemId id;
    private BookItemBarcode barcode;
    private IsReferenceOnly isReferenceOnly;
    private LoanCreationDate borrowedDate;
    private LoanDueDate dueDate;
    private Price price;
    private BookItemStatus status;
    private PurchaseDate dateOfPurchase;
    private BookId bookId;
    private RackId rackId;
    private ShelfId shelfId;
}
